Obsolete hash tracking

jagerman opened the pull request:
This adds tracking of obsolete hashes.  Doing so requires reworking some of the APIs around merging/pushing to now take the hash.

Fixes #19

jagerman committed:
Obsolete hash tracking

This adds tracking of obsolete hashes.  Doing so requires reworking some
of the APIs around merging/pushing to now take the hash.

jagerman committed:
Add missing C interface to legacy group members

jagerman commented:
This now also adds the legacy group member api that was missing from #15 (added here because the changes here add test suite differences).

mpretty-cyro reviewed (changes requested):
Looks like iOS is getting a build error (`Unknown type name 'bool'`) - need to add `#include <stdbool.h>` to the top of `community.h`

jagerman committed:
Add missing include for `bool`

jagerman committed:
Test suite fix for older gcc

jagerman committed:
Make current_hashes return a vector; expose in C API

`current_hash()` is currently just singular, but this will change in the
future when we add multi-part messages, so make it into a vector right
now.

Also map it in the C API.

jagerman committed:
Add partial URL (i.e. no-pubkey) URL parsing

This allows passing a base+room composed URL to `get_community` without
needing the specify the (unused) pubkey in the URL.  Session desktop, in
particular, requested this for easier details retrieval in some
contexts.

jagerman committed:
Remove unimplemented/unwanted code

This was left over from an initial version of the delete tracking that
ended up not being needed/wanted.

jagerman committed:
Formatting
